apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/infer-entities-from-web-page-worker-agent.ts

```diff
@@ -585,37 +667,69 @@ export const inferEntitiesFromWebPageWorkerAgent = async (params: {
               expectedNumberOfEntities,
               validAt,
               entityTypeIds: inferringEntitiesOfTypeIds,
+              linkEntityTypeIds: inferringLinkEntitiesOfTypeIds,
             } = toolCall.input as ToolCallArguments["inferEntitiesFromWebPage"];
 
-            const invalidEntityTypeIds = inferringEntitiesOfTypeIds.filter(
-              (entityTypeId) =>
-                !entityTypeIds.some(
-                  (expectedEntityTypeId) =>
-                    expectedEntityTypeId === entityTypeId,
-                ),
-            );
-
             if (expectedNumberOfEntities < 1) {
               return {
                 ...toolCall,
                 output: dedent(`
                   You provided an expected number of entities which is less than 1. You must provide
                   a positive integer as the expected number of entities to infer.
                 `),
+                isError: true,
               };
             }
 
-            if (invalidEntityTypeIds.length > 0) {
+            const validEntityTypeIds = input.entityTypes.map(({ $id }) => $id);
+
+            const invalidEntityTypeIds = inferringEntitiesOfTypeIds.filter(
+              (entityTypeId) => !validEntityTypeIds.includes(entityTypeId),
+            );
+
+            const validLinkEntityTypeIds = input.linkEntityTypes?.map(
+              ({ $id }) => $id,
+            );
+
+            const invalidLinkEntityTypeIds =
+              inferringLinkEntitiesOfTypeIds?.filter(
+                (entityTypeId) =>
+                  !validLinkEntityTypeIds?.includes(entityTypeId),
+              ) ?? [];
+
+            if (
+              invalidEntityTypeIds.length > 0 ||
+              invalidLinkEntityTypeIds.length > 0
+            ) {
               return {
                 ...toolCall,
                 output: dedent(`
-                  You provided invalid entity type IDs which don't correspond to the entity types
-                  which were initially provided: ${JSON.stringify(invalidEntityTypeIds)}
-
-                  The possible entity types you can submit are: ${JSON.stringify(
-                    entityTypeIds,
-                  )}
+                  ${
+                    invalidEntityTypeIds.length > 0
+                      ? dedent(`
+                        You provided invalid entity type IDs which don't correspond to the entity types
+                        which were initially provided: ${JSON.stringify(invalidEntityTypeIds)}
+
+                        The possible entity types you can submit are: ${JSON.stringify(
+                          validEntityTypeIds,
+                        )}
+                      `)
+                      : ""
+                  }
+                  ${
+                    invalidLinkEntityTypeIds.length > 0
+                      ? dedent(`
+                        You provided invalid link entity type IDs which don't correspond to the link entity types
```

Comment:
Quite possibly this isn't important, but I'd use the field name in the schemas when referring to inaccuracies, e.g. entityTypeIds rather than "entity type IDs". Might save errors at the margins and more consistency means less effort for us in parsing / naming at least

Reply:
That's a fair point, also not sure to what extent it would improve things but suspect it could be helpful. Have addressed this here for now in https://github.com/hashintel/hash/pull/4346/commits/f064a9d370157e0a000f0d5e7c6b3ef0818418f9, and keep this in mind when writing error messages for LLM consumption in future